### PR TITLE
ch-grow: fix COPY

### PR DESCRIPTION
**Pull request by reidpr**
Addresses #833 and #890.

I still need to add a little documentation but otherwise this is ready for review.

### CODE REVIEW

File: examples/copy/Dockerfile [hunk 1]
```diff
@@ -6,6 +6,7 @@
 # ch-test-builder-exclude: buildah-runc
 # ch-test-builder-exclude: buildah-setuid
 
+
```
Comment (j-ogas) [inline examples/copy/Dockerfile:9]: Extra newline?

File: examples/copy/Dockerfile [hunk 2]
```diff
@@ -145,6 +181,77 @@ RUN mkdir dir18
...
+RUN test $(stat -c '%A' dir19/dir19a2 | cut -c8-) = 'rwx' \
+ && stat -c '%n: %A' dir19/dir19a2
+COPY dirF dir19/
+RUN test $(stat -c '%A' dir19/dir19a2 | cut -c8-) != 'rwx' \
+ && stat -c '%n: %A' dir19/dir19a2
+
+
+## Destination: Symlink, 2nd level.
+
+# Note: This behavior is DIFFERENT from the symlink at 1st level tests above.
```
Comment (j-ogas) [inline examples/copy/Dockerfile:223]: Why? Perhaps a sentence or two of context?

File: lib/build.py [hunk 3]
```diff
@@ -353,13 +354,94 @@ def __init__(self, *args):
    def str_(self):
       return "%s -> %s" % (self.srcs, repr(self.dst))
 
+   def copy_src_dir(self, src, dst):
+      """Copy the contents of directory src, named by COPY, either explicitly
+         or with wildcards, to dst. src might be a symlink, but dst is a
+         canonical paths. Both must be at the top level of the COPY
```
Comment (j-ogas) [inline lib/build.py:360]:
```suggestion
         canonical path. Both must be at the top level of the COPY
```

File: lib/build.py [hunk 4]
```diff
@@ -353,13 +354,94 @@ def __init__(self, *args):
...
+               ch.DEBUG("destination exists, removing")
+               if (os.path.isdir(dst_path) and not os.path.islink(dst_path)):
+                  ch.rmtree(dst_path)
+               else:
+                  ch.unlink(dst_path)
+            ch.copy2(src_path, dst_path, follow_symlinks=False)
+
+   def copy_src_file(self, src, dst):
+      """Copy file src, named by COPY either explicitly or with wildcards, to
+         dst. src might be a symlink, but dst is a canonical paths. Both must
```
Comment (j-ogas) [inline lib/build.py:426]:
```suggestion
         dst. src might be a symlink, but dst is a canonical path. Both must
```

**Review by j-ogas — APPROVED** [review]

**Review by heasterday — COMMENTED** [review]
This looks good to me. That being said, it does not appear to fully resolve the issues building the lammps image, it fails with the following error:
```
 40 RUN ['/bin/sh', '-c', 'ldconfig']
 43 COPY ['melt.patch'] -> '/lammps/examples/melt/melt.patch'
error: can't COPY: destination not in image: /usr/local/src/lammps-patch_19Sep2019/examples/melt/melt.patch
```

**Review by heasterday — APPROVED** [review]

**Pull request merged**